glue/sales_val_flagging.py

```diff
@@ -131,6 +131,54 @@ def ptax_adjustment(df, groups, ptax_sd, condos: bool):
     directions = ["High", "Low"]
     df["ptax_direction"] = np.select(conditions, directions, default=np.nan)
 
+    # Utilize PTAX-203, complete binary columns
+    df = df.rename(columns={"sv_is_outlier": "sv_is_autoval_outlier"}).assign(
+        sv_is_autoval_outlier=lambda df: df["sv_is_autoval_outlier"] == 1,
+        sv_is_outlier=lambda df: df["sv_is_autoval_outlier"]
+        | df["ptax_flag_w_deviation"],
+    )
+
+    # Calculate the new values for sv_outlier_reason1 based on ptax conditions
+    new_sv_outlier_reason1 = np.select(
+        [
+            (df["ptax_flag_w_deviation"]) & (df["ptax_direction"] == "High"),
+            (df["ptax_flag_w_deviation"]) & (df["ptax_direction"] == "Low"),
+        ],
+        ["PTAX-203 Exclusion (High)", "PTAX-203 Exclusion (Low)"],
+        default=df["sv_outlier_reason1"],
+    )
+
+    # Conditionally shift sv_outlier_reason2 and sv_outlier_reason3
+    df = df.assign(
+        sv_outlier_reason2=lambda df: np.where(
+            (new_sv_outlier_reason1 != df["sv_outlier_reason1"])
+            & (df["sv_outlier_reason1"] != "Not outlier"),
+            df["sv_outlier_reason1"],
+            df["sv_outlier_reason2"],
+        ),
+        sv_outlier_reason3=lambda df: np.where(
+            (new_sv_outlier_reason1 != df["sv_outlier_reason1"])
+            & (df["sv_outlier_reason1"] != "Not outlier"),
+            df["sv_outlier_reason2"],
+            df["sv_outlier_reason3"],
+        ),
+    ).assign(
+        sv_outlier_reason1=new_sv_outlier_reason1  # Finally update sv_outlier_reason1 with the new values
+    )
+
+    df = df.assign(
+        # PTAX-203 binary
+        sv_is_ptax_outlier=lambda df: df["sv_outlier_reason1"].str.contains(
+            "PTAX-203", na=False
+        ),
+        # Heuristics flagging binary column
+        sv_is_heuristic_outlier=lambda df: (
+            ~df["sv_outlier_reason1"].str.contains("PTAX-203", na=False)
+        )
+        & (df["sv_is_outlier"] == 1),
+        sv_is_outlier=lambda df: df["sv_is_outlier"].astype(int),
+    )
+
```

Comment:
Two things are happening in this change.

1. We move some ptax processing code from `finish_flags()` into `ptax_adjustment()`. This improves organization - instead of ptax processing split across multiple functions, we now see most of the ptax processing under one function called `ptax_adjustment()`
2. We handle the ptax processing with the new design doc specifications. If the observation is a ptax outlier, then we assign it for the `sv_outlier_reason1`, shift the existing `sv_outlier_reason1` to `sv_outlier_reason2`, and shift the existing `sv_outlier_reason2` to `sv_outlier_reason3`.


First we take the old column and overwrite the `sv_outlier_reason1` if it is a ptax outlier.
```python
new_sv_outlier_reason1 = np.select(
        [
            (df["ptax_flag_w_deviation"]) & (df["ptax_direction"] == "High"),
            (df["ptax_flag_w_deviation"]) & (df["ptax_direction"] == "Low"),
        ],
        ["PTAX-203 Exclusion (High)", "PTAX-203 Exclusion (Low)"],
        default=df["sv_outlier_reason1"],
    )
```

Then if the following conditions are fulfilled (code and written)
```python
(new_sv_outlier_reason1 != df["sv_outlier_reason1"])
            & (df["sv_outlier_reason1"] != "Not outlier")
```
- the `new_sv_outlier_reason1` and the original `sv_outlier_reason1` aren't equal (due to ptax overwrite)
- the original `sv_outlier_reason1` isn't `Not outlier`

we shift the `sv_outlier_reason1` to `sv_outlier_reason2`. We do the same from `sv_outlier_reason2` to `sv_outlier_reason3`.